homeassistant/components/shelly/entity.py

```diff
@@ -449,13 +452,15 @@ class ShellyRestAttributeEntity(update_coordinator.CoordinatorEntity):
 
     def __init__(
         self,
-        wrapper: BlockDeviceWrapper,
+        wrapper: ShellyDeviceRestWrapper,
+        block_wrapper: BlockDeviceWrapper,
         attribute: str,
         description: RestEntityDescription,
     ) -> None:
         """Initialize sensor."""
         super().__init__(wrapper)
         self.wrapper = wrapper
+        self.block_wrapper = block_wrapper
```

Comment:
Do all RestAttribute entities need the block wrapper or is it only the update entity that needs that?

Reply:
Only the update entity needs it, but do mind that the block wrapper is setup anyway and is already present in the HomeAssistant data, this only links a self.block_wrapper variable to the wrapper in the HomeAssistant data.

Reply:
Looking at the code, the only reason `block_wrapper` is used here is because `async_trigger_ota_update` is implemented in the `BlockDeviceWrapper` class.

However this made me realize that if we need the `rest_wrapper` there is a change in the default integration behaviour which we missed during the review, the binary sensors were disabled by default since if they are enabled they introduce polling and there was a discussion in an older PR that we need to address this and poll update in lower rate, since what will happen now is that we pull all devices every 1 minute just to check for update.
This needs to be reverted.
Before:
https://github.com/home-assistant/core/blob/77a933d6f02439c713995d98712cc5b4513fd8fe/homeassistant/components/shelly/binary_sensor.py#L151
After:
https://github.com/home-assistant/core/blob/da445e515b77500ca80630540c0c0668f8af4a37/homeassistant/components/shelly/update.py#L73

Reply:
Can the polling for the update entities happen at a longer interval, say once per day?

Reply:
I guess you're saying this needs a larger refactor?

Reply:
> I guess you're saying this needs a larger refactor?

missed that comment, yes, not so big, but I think it is better not to do it during beta

Reply:
@thecode We can simply disable the entities by default and no polling will take place for the 2022.10 release.
However users are of course free to enable the update entities and normal users will have no idea about the polling rate or even that these update entities use polling.
So we schould fix this.

Reply:
Although if we disable the entities by default now, users that install 2022.10 will have the update entities created as disabled.
If we then fix the issue in 2022.11 and set the default to enabled, the already created entities in 2022.10 will keep beeing disabled.

So a user that is not aware of these new update entities will never benefit from them because the user will probably forget to enable it (after 2022.11)

Reply:
@thecode how long would it take you to implement slower polling for the update entities?
We are currently only in the second day of beta...

(for me it will probably take much longer since I am not familiar with the shelly code prior to this PR).

Reply:
> Do all RestAttribute entities need the block wrapper or is it only the update entity that needs that?

While working on https://github.com/home-assistant/core/pull/79628 and adding typing, I realized that there is a basic design problem in our REST coordinator and another typing problem in https://github.com/home-assistant/core/pull/78305 which leaded to this PR, the REST coordinator is used to pull data via REST for Gen1 devices to fetch diagnostics data that is not pushed by the device, however the data itself stored in the `BlockDevice` class and effectively we don't really have two places that holds data (the REST coordinator does not store data it just trigger polling). Other places that used the REST coordinator also fetch the data from the `BlockDevice` and use the Block coordinator typing. I have fixed that also for the `update` platform so it doesn't have two coordinators/wrappers.

I am closing this PR. 

As for the support for polling at a lower rate, we currently have 4 coordinators (wrongly named wrappers), I plan to merge them and leave just two, 1 for block based devices (CoAP, Gen1) and one for RPC based devices (WS, Gen2)

Existing Gen1:
- `BlockDeviceWrapper` -> `ShellyBlockCoordinator`
- `ShellyDeviceRestWrapper` -> `ShellyRestCoordinator`

Existing Gen2:
- `RpcPollingWrapper` -> `ShellyRpcPollingCoordinator`
- `RpcDeviceWrapper` -> `ShellyRpcCoordinator`

So I plan to keep only the  `ShellyBlockCoordinator` and `ShellyRpcCoordinator` with an attribute that trigger polling for certain entities if they are enabled, I will also try to address the request here and add an attribute to poll in a lower rate for update entites.